rust-runtime/aws-smithy-http-server-python/examples/stubgen.py

```diff
@@ -0,0 +1,391 @@
+        filter(lambda n: n != "object", map(lambda b: b.__name__, klass.__bases__))
+    )
+
+    class_sig = DocstringParser.parse_class(klass)
+    if class_sig:
+        (generics, extends) = class_sig
+        base_classes.extend(map(writer.fix_and_include, extends))
+        for g in generics:
+            writer.generic(g)
+
+    bases = ", ".join(base_classes)
+    definition = f"class {class_name}({bases}):\n"
+
+    def preserve_doc(obj: Any) -> str:
+        return format_doc(obj, indent_level + 4) + "\n"
+
+    definition += preserve_doc(klass)
+
+    is_empty = True
+    klass_vars: Dict[str, Any] = vars(klass)
+    for name, member in sorted(klass_vars.items(), key=lambda k: k[0]):
+        if name.startswith("__"):
+            continue
+
+        if inspect.isdatadescriptor(member):
+            is_empty = False
+            definition += (
+                indent(make_field(writer, name, member), indent_level + 4) + "\n"
+            )
+            definition += preserve_doc(member)
+        elif is_fn_like(member):
+            is_empty = False
+            definition += make_function(writer, name, member, indent_level + 4) + "\n"
+        # Enum variant
+        elif isinstance(member, klass):
+            is_empty = False
+            definition += indent(f"{name}: {class_name}\n", indent_level + 4)
+            definition += preserve_doc(member)
+        else:
+            print(f"Unknown member type={member}")
+
+    if inspect.getdoc(klass) is not None:
+        constructor_sig = DocstringParser.parse(klass)
+        if constructor_sig is not None and (
+            # Make sure to only generate `__init__` if the class has a constructor defined
+            len(constructor_sig.rtypes) > 0
+            or len(constructor_sig.params) > 0
+        ):
+            is_empty = False
+            definition += (
+                make_function(
+                    writer, "__init__", klass, indent_level + 4, include_docs=False
+                )
+                + "\n"
+            )
+
+    if is_empty:
+        definition += indent(f"...\n", indent_level + 4)
+
+    return definition
+
+
+def walk_module(writer: Writer, mod: Any):
+    exported = mod.__all__
+
+    for (name, member) in inspect.getmembers(mod):
+        if name not in exported:
+            continue
+
+        if inspect.ismodule(member):
+            walk_module(
+                writer.submodule(writer.path.parent / name / "__init__.pyi"), member
+            )
+        elif inspect.isclass(member):
+            writer.define(make_class(writer, name, member))
+        elif is_fn_like(member):
+            writer.define(make_function(writer, name, member, free_standing=True))
+        else:
+            print(f"Unknown type: {member}")
+
+
+if __name__ == "__main__":
+    import argparse
+    import importlib
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("module")
+    parser.add_argument("outdir")
+    args = parser.parse_args()
```

Comment:
This code is quite complex and hard to review (it's a hard problem) :D I would like to see some tests in this file to ensure we are doing the right thing and we do not regress in the future.

I think you have several options:
* Make a test suite using pytest with a standard python module
* Add an option to the main, something like `--run-tests`, which will lazily import `pytest` and run something like `pytest.run()`, which will collect the test cases and run them
* Delegate the testing to the github CI or to something written in kotlin in the code generator.

What do you think?

Reply:
I don't think tests are **a requirement** for this PR, but we should work on them before releasing this feature as stable.